dockerHashing/src/main/scala/cromwell/docker/registryv2/flows/azure/AzureContainerRegistry.scala

```diff
@@ -0,0 +1,149 @@
+    dockerImageIdentifier.hostAsString.contains(domain) 
+
+  override protected def authorizationServerHostName(dockerImageIdentifier: DockerImageIdentifier): String =
+    dockerImageIdentifier.host.getOrElse("")
+
+  /**
+    * In Azure, service name does not exist at the registry level, it varies per repo, e.g. `terrabatchdev.azurecr.io`
+    */
+  override def serviceName: Option[String] =
+    throw new Exception("ACR service name is host of user-defined registry, must derive from `DockerImageIdentifier`")
+  
+  /**
+    * Builds the list of headers for the token request
+    */
+  override protected def buildTokenRequestHeaders(dockerInfoContext: DockerInfoContext): List[Header] = {
+    List(contentTypeHeader)
+  }
+
+  private val contentTypeHeader: Header = {
+    import org.http4s.headers.`Content-Type`
+    import org.http4s.MediaType
+
+    `Content-Type`(MediaType.application.`x-www-form-urlencoded`)
+  }
+  
+  private def getRefreshToken(authServerHostname: String, defaultAccessToken: String): IO[Request[IO]] = {
+    import org.http4s.Uri.{Authority, Scheme}
+    import org.http4s.client.dsl.io._
+    import org.http4s._
+
+    val uri = Uri.apply(
+      scheme = Option(Scheme.https),
+      authority = Option(Authority(host = Uri.RegName(authServerHostname))),
+      path = "/oauth2/exchange",
+      query = Query.empty
+    )
+
+    org.http4s.Method.POST(
+      UrlForm(
+        "service" -> authServerHostname,
+        "access_token" -> defaultAccessToken,
+        "grant_type" -> "access_token"
+      ),
+      uri,
+      List(contentTypeHeader): _* 
+    )
+  }
+
+  /*
+  Unlike other repositories, Azure reserves `GET /oauth2/token` for Basic Authentication [0]
+  In order to use Oauth we must `POST /oauth2/token` [1]
+  
+  [0] https://github.com/Azure/acr/blob/main/docs/Token-BasicAuth.md#using-the-token-api
+  [1] https://github.com/Azure/acr/blob/main/docs/AAD-OAuth.md#calling-post-oauth2token-to-get-an-acr-access-token 
+   */
+  private def getDockerAccessToken(hostname: String, repository: String, refreshToken: String): IO[Request[IO]] = {
+    import org.http4s.Uri.{Authority, Scheme}
+    import org.http4s.client.dsl.io._
+    import org.http4s._
+
+    val uri = Uri.apply(
+      scheme = Option(Scheme.https),
+      authority = Option(Authority(host = Uri.RegName(hostname))),
+      path = "/oauth2/token",
+      query = Query.empty
+    )
+
+    org.http4s.Method.POST(
+      UrlForm(
+        // Tricky behavior - invalid `repository` values return a 200 with a valid-looking token.
+        // However, the token will cause 401s on all subsequent requests.
+        "scope" -> s"repository:$repository:pull",
+        "service" -> hostname,
+        "refresh_token" -> refreshToken,
+        "grant_type" -> "refresh_token"
+      ),
+      uri,
+      List(contentTypeHeader): _* // http4s adds `Content-Length` which ACR does not like (400 response) 
```

Comment:
How does this syntax prevent the Content-Length header from being added?

Reply:
Apologies, this is a comment based on old info. Removing.